src/dune/sub_dirs.ml

```diff
@@ -112,34 +130,86 @@ let eval (t : _ Status.Map.t) ~dirs =
                 dir
             ]))
 
-let decode =
+type subdir_stanzas = (Loc.t * Predicate_lang.Glob.t) option Status.Map.t
+
+module Dir_map = struct
+  type per_dir =
+    { sexps : Dune_lang.Ast.t list
+    ; subdir_status : subdir_stanzas
+    }
+
+  type t =
+    { data : per_dir
+    ; nodes : t String.Map.t
+    }
+
+  let root t = t.data
+
+  let descend t (p : string) = String.Map.find t.nodes p
+
+  let sub_dirs t = String.Map.keys t.nodes
+
+  let merge_data d1 d2 =
+    { sexps = d1.sexps @ d2.sexps
```

Comment:
Technically, the performance of the old implementation was better. Now we are quadratic in the number of directories contributing stanzas. I don't think this is a big deal however.